airflow/runtimes/podman_runtime.go

```diff
@@ -1,10 +1,322 @@
+		if err != nil {
+			return err
+		}
+
+		// If the machine is already running,
+		// just go ahead and configure it for usage.
+		if iMachine.State == podmanStatusRunning {
+			return rt.configureMachineForUsage(iMachine)
+		}
+
+		// If the machine is stopped,
+		// start it, then configure it for usage.
+		if iMachine.State == podmanStatusStopped {
+			s.Start()
+			if err := rt.Engine.StartMachine(podmanMachineName); err != nil {
+				return err
+			}
+			return rt.configureMachineForUsage(iMachine)
+		}
+	}
+
+	// Otherwise, initialize the machine
+	s.Start()
+	if err := rt.Engine.InitializeMachine(podmanMachineName); err != nil {
+		return err
+	}
+
+	return rt.getAndConfigureMachineForUsage(podmanMachineName)
+}
+
+// stopAndKillMachine attempts to stop and kill the Podman machine.
+// If other projects are running, it will leave the machine up.
+func (rt PodmanRuntime) stopAndKillMachine() error {
+	// If the machine doesn't exist, exist early.
+	if !rt.astroMachineExists() {
+		return nil
+	}
+
+	// If the machine exists, and its running, we need to check
+	// if any other projects are running. If other projects are running,
+	// we'll leave the machine up, otherwise we stop and kill it.
+	if rt.astroMachineIsRunning() {
+		// Get the containers that are running on our machine.
+		containers, err := rt.Engine.ListContainers()
+		if err != nil {
+			return err
+		}
+
+		// Check the container labels to identify if other projects are running.
+		projectNames := make(map[string]struct{})
+		for _, item := range containers {
+			// Check if "project.name" exists in the Labels map
+			if projectName, exists := item.Labels[composeProjectLabel]; exists {
+				// Add the project name to the map (map keys are unique)
+				projectNames[projectName] = struct{}{}
+			}
+		}
+
+		// At this point in the command hook lifecycle, our project has already been stopped,
+		// and we are checking to see if any additional projects are running.
+		if len(projectNames) > 0 {
+			return nil
+		}
+
+		// If we made it this far, we can stop the machine,
+		// as there are no more projects running.
+		err = rt.Engine.StopMachine(podmanMachineName)
+		if err != nil {
+			return err
+		}
+	}
+
+	// If we make it here, the machine was already stopped, or was just stopped above.
+	// We can now remove it.
+	err := rt.Engine.RemoveMachine(podmanMachineName)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// configureMachineForUsage does two things:
+//   - Sets the DOCKER_HOST environment variable to the machine's socket path
+//     This allows the docker compose library to function as expected.
+//   - Sets the podman default connection to the machine
+//     This allows the podman command to function as expected.
+func (rt PodmanRuntime) configureMachineForUsage(machine *types.InspectedMachine) error {
+	if machine == nil {
+		return fmt.Errorf("machine does not exist")
+	}
+
+	if !rt.OSChecker.IsWindows() {
```

Comment:
@pritt20 how does this work on Windows if we don't set the `DOCKER_HOST`? How does the compose library communicate with podman?

Reply:
Hey @schnie ,

In windows podman uses `npipe` to set connection path in DOCKER_HOST environment variable and by default it point to docker engine `npipe:////./pipe/docker_engine` , therefore we do not require to set this explicitly. Also, unlike mac, we would look into `.ConnectionInfo.PodmanPipe.Path` in machine inspect to get this connection information.

Reference doc: https://podman-desktop.io/docs/migrating-from-docker/using-the-docker_host-environment-variable

Since in windows podman uses WSL it requires ssh connection to access the WSL linux vm and therefore podman command look for `.ssh/known_host` file to see if podman machine host is accessible.